build_tools/github_actions/runner/instance_deleter/main.py

```diff
@@ -0,0 +1,211 @@
+        request (flask.Request): The request object.
+        <https://flask.palletsprojects.com/en/1.1.x/api/#incoming-request-data>
+    Returns:
+        The response text, or any set of values that can be turned into a
+        Response object using `make_response`
+        <https://flask.palletsprojects.com/en/1.1.x/api/#flask.make_response>.
+    Note:
+        For more information on how Flask integrates with Cloud
+        Functions, see the `Writing HTTP functions` page.
+        <https://cloud.google.com/functions/docs/writing/http#http_frameworks>
+  """
+  if request.method != "DELETE":
+    return flask.abort(
+        400, f"Invalid method '{request.method}'. Only DELETE is supported.")
+
+  # No path is needed, since the token contains all the information we need.
+  if request.path != "/":
+    return flask.abort(
+        400,
+        f"Invalid request path '{request.path}'. Only root path is valid).")
+
+  auth_header = request.headers.get("Authorization")
+  if auth_header is None:
+    return flask.abort(401, "Authorization header is missing")
+  if not auth_header.startswith(AUTH_HEADER_PREFIX):
+    return flask.abort(
+        401, f"Authorization header does not start with expected string"
+        f" '{AUTH_HEADER_PREFIX}'.")
+
+  token = auth_header[len(AUTH_HEADER_PREFIX):]
+
+  try:
+    # We don't verify audience here because Cloud IAM will have already done so
+    # and jwt's matching of audiences is exact, which means trailing slashes or
+    # http vs https matters and that's pretty brittle.
+    token_payload = verify_token(token)
+  except ValueError as e:
+    print(e)
+    return flask.abort(401, "Decoding bearer token failed.")
+
+  print(f"Token payload: {token_payload}")
+
+  try:
+    compute_info = token_payload["google"]["compute_engine"]
+  except KeyError as e:
+    return flask.abort(
+        401,
+        "Bearer token payload does not have expected field 'google.compute'")
+
+  project = compute_info["project_id"]
+  zone = compute_info["zone"]
+  instance_name = compute_info["instance_name"]
+  try:
+    instance = instances_client.get(instance=instance_name,
+                                    project=project,
+                                    zone=zone)
+  except exceptions.NotFound as e:
+    print(e)
+    return flask.abort(404, "Instance not found")
+
+  instance_id = int(compute_info["instance_id"])
+  # Verify it's *actually* the same instance. Names get reused, but IDs
+  # don't. For some reason you can't reference instances by their ID in any
+  # of the APIs.
+  if instance.id != instance_id:
+    return flask.abort(
+        400,
+        f"Existing instance of the same name '{instance.name}' has a different"
+        f" ID '{instance.id}' than token specifies '{instance_id}'.")
+
+  mig = get_from_items(instance.metadata.items, "created-by")
+
+  if mig is None:
+    return flask.abort(400, "Instance is not part of a managed instance group.")
+  mig = get_name_from_resource(mig)
```

Comment:
Wondering if we should have an allow list of MIGs? Otherwise this looks like we are creating an API that allows any program on any instance in MIG to delete itself (regardless of the Linux user accounts IIUC), as long as the cloud run has permissions on the MIGs.

This might not be intended for all MIGs in the future? (Don't have a concrete example here, but it seems to be easy to forget we have this cloud run that provides such mechanisms)

Reply:
That can be controlled via the IAM of the Cloud Functions/Run service account. We'll only give it permissions to the MIGs for which we want to enable this functionality.

Reply:
Oh nevermind. It turns out that you can't set per-resource IAM on instance groups or have conditional rules that reference them :upside_down_face:. I guess a manual list it is...